File: solver.py
import sys

WORDFILE = "SOWPODS.txt"
MAX = 15 # longest word in sowpods is 5
BLANK = '?'
blanks = 0

def solve(letters, prefix='', suffix=''):
    matches = [] # store the results here

    # extract blanks
    global blanks 
    blanks = letters.count(BLANK)
    if blanks > 0:
        letters = letters.replace(BLANK, '')

    with open(WORDFILE) as f:
        for line in f:
            w = str(line).rstrip()
            if ( len(w) > len(letters) + len(prefix) + len(suffix) or
                 len(w) > MAX ): break

            word = word_matches(w, letters, prefix, suffix)
            if word:
                matches.append(word)

    return matches

# return false if: w doesn't start/end with prefix/suffix,
# or if w cannot be formed from letters
def word_matches(w, letters, prefix, suffix):
    word = w # make a copy of word

    if prefix: # if prefix is blank don't do anything with ti
        if w.startswith(prefix):
            w = w.replace(prefix, '', 1) # delete prefix from w
        else: 
            return False

    if suffix:
        if w.endswith(suffix):
            w = w.replace(suffix, '', 1)
        else: 
            return False

    global blanks
    for c in w:
        if c not in letters: 
            if blanks > 0:
                blanks -= 1
                # change word here?
            else:
                return False
        else:
            letters = letters.replace(c, '', 1)

    return word

if __name__ == "__main__": # if run from command line (vs imported)
    if len(sys.argv) > 1: 
        letters = sys.argv[1] 
    else: 
        print("give me some letters")
        quit()

    prefix = sys.argv[2] if len(sys.argv) > 2 else ''
    suffix = sys.argv[3] if len(sys.argv) > 3 else ''
    print(prefix)
    print(suffix)
    print("finding words for %s..." % letters)
    r = solve(letters, prefix, suffix)
    print("%s\n\nI found %s words" % (r, len(r)))

File: local-only/solver.js
const Solver = () => {
  //const DICT = "http://10.0.1.5:8000/local-only/SOWPODS.txt"
  const DICT = "http://cyberpixels.net/anagram/SOWPODS.txt"
  const api = {
    solve: solve
  }

  let dict = undefined
  fetch(DICT)
    .then(response => response.text())
    .then(text => dict = text.split('\n'))

  function solve(letters, prefix, suffix) {
    if(dict === undefined) {
      console.log("still loading the dictionary")
      return
    }

    let matches = []
    
    for(let i = 0; i < dict.length; i++) {
      if(dict[i].length > letters.length + prefix.length + suffix.length) {
        break
      }

      if(wordMatches(dict[i], letters, prefix, suffix)) {
        matches.push(dict[i])
      }
    }

    console.log(matches.length + " words found")
    return matches
  }

  function isCapital(l) {
    return l !== l.toLowerCase()
  }

  function wordMatches(word, letters, prefix, suffix) {
    if(prefix && prefix.length > 0) {
      if(word.startsWith(prefix)) {
        word = word.slice(prefix.length)
      } else {
        return false
      }
    }

    if(suffix && suffix.length > 0) {
      if(word.endsWith(suffix)) {
        word = word.slice(0, word.length - suffix.length)
      } else {
        return false
      }
    }

    for(let i = 0; i < letters.length; i++) {
      if(isCapital(letters[i])) {
        if(word[i] !== letters[i].toLowerCase()) return false
      }
    }

    let numblanks = letters.split('?').length - 1
    letters = letters.toLowerCase().replace('?', '').split('')

    for(let i = 0; i < word.length; i++) {
      let n = letters.indexOf(word[i])
      if(n < 0) {
        if(numblanks > 0) {
          numblanks--
        } else {
          return false
        }
      }
      letters[n] = null
    }

    return true
  }

  return api
}
